zap/src/main/java/org/zaproxy/zap/extension/ascan/filters/GenericFilterUtility.java

```diff
@@ -0,0 +1,272 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2019 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.ascan.filters;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
+import org.apache.commons.collections.CollectionUtils;
+import org.parosproxy.paros.Constant;
+import org.zaproxy.zap.extension.ascan.filters.impl.AlertScanFilter;
+import org.zaproxy.zap.extension.ascan.filters.impl.ConfidenceLevelScanFilter;
+import org.zaproxy.zap.extension.ascan.filters.impl.HttpStatusCodeScanFilter;
+import org.zaproxy.zap.extension.ascan.filters.impl.MethodScanFilter;
+import org.zaproxy.zap.extension.ascan.filters.impl.TagScanFilter;
+import org.zaproxy.zap.extension.ascan.filters.impl.URLPatternScanFilter;
+
+/**
+ * Generic Filter Utility Class.
+ *
+ * @author KSASAN preetkaran20@gmail.com
+ */
+public class GenericFilterUtility {
+
+    public static final String INCLUDE_FILTER_CRITERIA_MESSAGE_KEY =
+            "scan.filter.filtercriteria.include";
+    public static final String EXCLUDE_FILTER_CRITERIA_MESSAGE_KEY =
+            "scan.filter.filtercriteria.exclude";
+    public static final String INCLUDE_ALL_FILTER_CRITERIA_MESSAGE_KEY =
+            "scan.filter.filtercriteria.includeall";
+
+    /**
+     * this Utility is Generic Utility for various criteria's and can be used by any Filter which is
+     * based on classes which are implementing hashCode and equals
+     *
+     * @param <T> any object implementing hashCode and equals method
+     * @param genericFilterDataCollection Collection of data provided by setting filter criteria
+     * @param nodeValues data associated with the scanned node.
+     * @param filterType type of the filter ie Tag/Http Status code/Method etc
+     * @return FilterResult contains result after applying filter.
+     */
+    public static <T> FilterResult isFiltered(
```

Comment:
As i thought it as a utility and can be used at other places too and doesn't require each class to extend the AbstractScanFilter. Also as you suggested to add the Criteria to Filter, so it can be made more generic and reusable.

Reply:
Which other places? ;) This is tied to active scanner so we would have to change it further to support other places.

Reply:
not sure but it can be utilised in passive scanner too, may be in future.

Reply:
Well, we can worry about supporting those places if/when that happens, it's clear that this utility would not be used (as is per depending on active scan classes), so there's no need to try support those places now.

Reply:
Last point, but now it will not depend on active scan classes except filter criteria as i will remove GenericFilterData as you suggested as it is having one collection and FilterCriteria but now i will pass both the data to this method separately. 

if you think still think adding this to Abstract class, i will do that.

Reply:
The result is still an active scan class (all are in active scan package).

> as you suggested

Note that I was not suggesting that, just discussing possibilities.

Reply:
> The result is still an active scan class (all are in active scan package).
> 
> > as you suggested
> 
> Note that I was not suggesting that, just discussing possibilities.

then i will add to the Abstract class.